src/protocols/secure_channel/CASESession.cpp

```diff
@@ -498,114 +461,103 @@ CHIP_ERROR CASESession::SendSigmaR2()
         SuccessOrExit(err = tlvWriter.Finalize());
+        msg_r2_signed_enc_len = static_cast<uint16_t>(tlvWriter.GetLengthWritten());
     }
 
-    // Step 9
-    err = AES_CCM_encrypt(msg_R2_Encrypted.Get(), msg_r2_signed_enc_len, nullptr, 0, sr2k, kAEADKeySize, kIVSR2, kIVLength,
-                          msg_R2_Encrypted.Get(), tag, sizeof(tag));
+    // Generate the encrypted data blob
+    err = AES_CCM_encrypt(msg_R2_Encrypted.Get(), msg_r2_signed_enc_len, nullptr, 0, sr2k, kAEADKeySize, kTBEData2_Nonce,
+                          kTBEDataNonceLength, msg_R2_Encrypted.Get(), tag, sizeof(tag));
     SuccessOrExit(err);
 
-    data_len = static_cast<uint16_t>(kSigmaParamRandomNumberSize + sizeof(uint16_t) + kTrustedRootIdSize + kP256_PublicKey_Length +
-                                     msg_r2_signed_enc_len + sizeof(tag) + sizeof(uint64_t) * 6);
+    // Construct Sigma2 Msg
+    data_len = static_cast<uint16_t>(kSigmaParamRandomNumberSize + sizeof(uint16_t) + kP256_PublicKey_Length +
+                                     msg_r2_signed_enc_len + sizeof(tag) + sizeof(uint64_t) * 5);
 
     msg_R2 = System::PacketBufferHandle::New(data_len);
     VerifyOrExit(!msg_R2.IsNull(), err = CHIP_ERROR_NO_MEMORY);
 
-    // Step 10
-    // now construct sigmaR2
     {
         System::PacketBufferTLVWriter tlvWriter;
         TLV::TLVType outerContainerType = TLV::kTLVType_NotSpecified;
 
         tlvWriter.Init(std::move(msg_R2));
         SuccessOrExit(err = tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
-        SuccessOrExit(err = tlvWriter.PutBytes(CASETLVTag::kRandom, msg_rand.Get(), kSigmaParamRandomNumberSize));
-        SuccessOrExit(err = tlvWriter.Put(CASETLVTag::kSessionID, mConnectionState.GetLocalKeyID(), true));
-        SuccessOrExit(err = tlvWriter.PutBytes(CASETLVTag::kTrustedRootID, mTrustedRootId.data(),
-                                               static_cast<uint32_t>(mTrustedRootId.size())));
-        SuccessOrExit(err = tlvWriter.PutBytes(CASETLVTag::kResponderEphPubKey, mEphemeralKey.Pubkey(),
+        SuccessOrExit(err = tlvWriter.PutBytes(TLV::ContextTag(1), msg_rand.Get(), kSigmaParamRandomNumberSize));
+        SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(2), mConnectionState.GetLocalKeyID(), true));
+        SuccessOrExit(err = tlvWriter.PutBytes(TLV::ContextTag(3), mEphemeralKey.Pubkey(),
                                                static_cast<uint32_t>(mEphemeralKey.Pubkey().Length())));
-        SuccessOrExit(err = tlvWriter.PutBytes(CASETLVTag::kEncryptedData, msg_R2_Encrypted.Get(), msg_r2_signed_enc_len));
-        SuccessOrExit(err = tlvWriter.PutBytes(CASETLVTag::kTag, tag, sizeof(tag)));
+        SuccessOrExit(err = tlvWriter.PutBytes(TLV::ContextTag(4), msg_R2_Encrypted.Get(), msg_r2_signed_enc_len));
+        SuccessOrExit(err = tlvWriter.PutBytes(TLV::ContextTag(5), tag, sizeof(tag)));
```

Comment:
Tag 5 does not exist. The "tag" (i.e. message tag from  from `AES_CCM_encrypt` must be concatenated to the encrypted data.

The crypto primitive in spec:
```
byte[lengthInBytes(P) + CHIP_CRYPTO_AEAD_MIC_LENGTH_BYTES]
CHIP_Crypto_AEAD_GenerateEncrypt(
     SymmetricKey K,
     byte[lengthInBytes(P)] P,
     byte[] A,
     byte[CHIP_CRYPTO_AEAD_NONCE_LENGTH_BYTES] N)
```

The buffer for msg_R2_Encrypted  should have been allocated with at least kAEADKeySize extra, and then pointed the tag buffer to `msg_R2_Encrypted.Get() + msg_r2_signed_enc_len` so that the tag is emitted at the end, rather than in a separate buffer.